hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java

```diff
@@ -1085,13 +1086,14 @@ public static void checkMappedByType(
 			toOne = (ToOne) targetValue;
 		}
 		final String referencedEntityName = toOne.getReferencedEntityName();
-		PersistentClass referencedClass = propertyHolder.getPersistentClass();
-		while ( referencedClass != null ) {
-			if ( referencedClass.getEntityName().equals( referencedEntityName ) ) {
+		final PersistentClass referencedClass = persistentClasses.get( referencedEntityName );
+		PersistentClass ownerClass = propertyHolder.getPersistentClass();
+		while ( ownerClass != null ) {
+			if ( checkReferencedClass( ownerClass, referencedClass ) ) {
 				return;
 			}
 			else {
-				referencedClass = referencedClass.getSuperclass();
+				ownerClass = ownerClass.getSuperclass();
```

Comment:
Theoretically, a `PersistentClass` could have a `MappedSuperclass` instead which then extends from a `PersistentClass`, so it might be worth using a similar approach as in `org.hibernate.metamodel.internal.MetadataContext#applyGenericEmbeddableProperties` to traverse the super types.

Reply:
Added similar logic to get the superclass.